### PR TITLE
WIP: `Trainer` changes for distributed training

### PR DESCRIPTION
**Pull request by scarecrow1123**
Followup PR to #3390 and #3372 to bring in distributed training support. Following are the major changes done:

* Workers are spawned using `mp.spawn` and each worker creates its own `Trainer` instance
* `Trainer.__init__` wraps up `self.model` with `DistributedDataParallel`
*  Logging and metric aggregation are already done in the previous PRs
* `Vocabulary` creation in case of distributed training is done before spawning the workers and creating `Trainer` class

To run distributed training, the trainer needs to have the following flag to be enabled:

```jsonnet
{
    "trainer": {
        "distributed": true,
        // ...
    }
}
```

TODO:
* Try to reproduce comparable results and share extensive results for existing/selected models
* Check if other commands like `evaluate`, `predict`, `fine-tune` works well with the new changes
* Should all the callbacks need to be called from every worker in case callback based training?
* Should the current dataset readers be changed to support distributed training as well?(to selectively yield data based on their rank)
* Write tests - _would be happy to get some suggestions on how to write tests for this_

@DeNeutoy

### CODE REVIEW

File: allennlp/models/biattentive_classification_network.py [hunk 1]
```diff
@@ -129,13 +129,10 @@ def __init__(
...
-                    "match the number of use_*_elmo flags set to true. use_input_elmo "
-                    "is %s, and use_integrator_output_elmo is %s".format(
-                        str(len(self._elmo._scalar_mixes)),
-                        str(self._use_input_elmo),
-                        str(self._use_integrator_output_elmo),
-                    )
+                    f"Elmo object has num_output_representations={len(self._elmo._scalar_mixes)}, but this "
+                    f"does not match the number of use_*_elmo flags set to true. use_input_elmo "
+                    f"is {self._use_input_elmo}, and use_integrator_output_elmo "
+                    f"is {self._use_integrator_output_elmo}"
```
[the remark this answers is not in the document]
Comment (scarecrow1123) [reply allennlp/models/biattentive_classification_network.py]: Yes this was enforced by `black` and hence the change to pass the CI.

File: allennlp/training/callback_trainer.py [hunk 2]
```diff
@@ -146,6 +148,18 @@ def __init__(
         # For capturing errors that occur during the train loop.
         self.exception: Optional[Exception] = None
 
+        # Using `DistributedDataParallel`(ddp) brings in a quirk wrt AllenNLP's `Model` interface and its
+        # usage. A `Model` object is wrapped by `ddp`, but assigning the wrapped model to `self.model`
+        # will break the usages such as `Model.get_regularization_penalty`, `Model.get_metrics`, etc.
```
[the remark this answers is not in the document]
Comment (scarecrow1123) [reply allennlp/training/callback_trainer.py]: > Have you and @DeNeutoy talked much about how metrics will work across processes? 

Metrics aggregation across processes is handled in `training_util.get_metrics` already which has been done in #3372 . So that part should just be fine for now I guess.

> The regularization penalty also seems like it could be an issue...

I'm not sure if that would be an issue as the penalty is computed from the `named_parameters` of the model. Since the gradients are synced across workers before the optimization step, this might just work well as is. Please correct me if I'm wrong. 

The reason I added the said comment is primarily to highlight the inconsistencies that may arise if we wrap `allennlp.model.Model` with `DistributedDataParallel` and use the same object throughout the trainer. Wrapping is necessary and it would break the `Model` specific interface methods such as `get_regularization_penalty`, etc. Hence the wrapped reference with a different name (`self._pytorch_model`) will only be used for pytorch specific ops such as model forward, etc. and the usages will only be through the actual `Model` instance.

File: allennlp/training/trainer_pieces.py [hunk 3]
```diff
@@ -78,7 +80,9 @@ def from_params(
         model.extend_embedder_vocab()
 
         # Initializing the model can have side effect of expanding the vocabulary
-        vocab.save_to_files(os.path.join(serialization_dir, "vocabulary"))
+        # Save the vocab only in the master
+        if not dist.is_initialized() or dist.get_rank() == 0:
+            vocab.save_to_files(os.path.join(serialization_dir, "vocabulary"))
```
[the remark this answers is not in the document]
Comment (scarecrow1123) [reply allennlp/training/trainer_pieces.py]: This needs to be done in case of non-distributed training as the vocab creation wouldn't be done up until this point. Whereas in the distributed case, it is done before spawning the workers since the entire view of the data is not available to the workers.

File: allennlp/commands/train.py [hunk 4]
```diff
@@ -268,11 +270,48 @@ def train_model(
...
+        device_id = parse_cuda_device(cuda_device)
+
+        if not isinstance(device_id, list):
+            raise ConfigurationError(
+                "Multiple cuda devices need to be configured to run distributed training."
+            )
+
+        logging.info("Switching to distributed training mode since multiple GPUs are configured")
+        os.environ["MASTER_ADDR"] = "127.0.0.1"
+        os.environ["MASTER_PORT"] = "29500"
```
[the remark this answers is not in the document]
Comment (scarecrow1123) [reply allennlp/commands/train.py]: The reason I did not have this configurable was the assumption that this will only apply to single-node multi-GPU training. These properties will only have significance in a multi-node setup and I did not have such a setup to test till last week. But I was able to run distributed training a multi-node setup with only a few more changes. Hence this would be changed to be configurable.

**Comment by scarecrow1123** [comment]
@brendan-ai2 Thank you for the suggestions and apologies for the delay in my response. I'm to yet start working on the tests. As I've mentioned in one of the above replies, I've also done a few changes to make the distributed training work in a multi-node multi-GPU setup. Adding those commits as well.

File: allennlp/models/biattentive_classification_network.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (scarecrow1123) [reply allennlp/models/biattentive_classification_network.py]: Yes, this was enforced by `black` and hence the change to pass the CI.

File: allennlp/training/trainer_pieces.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (brendan-ai2) [reply allennlp/training/trainer_pieces.py]: Ah, sorry. I think I was confused as to the original purpose of "# Initializing the model can have side effect of expanding the vocabulary", but that makes sense now.

File: allennlp/training/callback_trainer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brendan-ai2) [reply allennlp/training/callback_trainer.py]: I think this makes sense. Thanks for explaining!

**Comment by DeNeutoy** [comment]
Oh sorry I missed @brendan-ai2's note about adding tests to this PR. This seems like a good idea, I forgot we had tests for the multi-gpu stuff which we run occasionally already.

**Comment by brendan-ai2** [comment]
Yeah, if we can at least test the single-node stuff in this PR, I'll be happy :)

**Comment by brendan-ai2** [comment]
Hi @scarecrow1123, just wanted to check in on this PR. In case you're busy with other things, @DeNeutoy and I were discussing merging this into the branch as-is and then we could work on some of the tests ourselves. Definitely don't want to step on your toes, but I'd love to keep the momentum here going. Thanks again for all your work on this! :)

**Comment by scarecrow1123** [comment]
Hey @brendan-ai2  I had planned to bring this PR to a closure over the last weekend and I couldn't do that. Apologies for dragging this up a bit. (_Basically it is the end of fall term in grad school + full time work ;)_ ) I'll work on this the coming weekend, but if you guys think it would be better to merge it to the branch before that, please go ahead. Either of the options would be good as anyways I'll do a catch up with more tests myself. Thanks for the nudge!

**Comment by brendan-ai2** [comment]
Oof, grad school + full time work is a lot! Thanks again for finding time to work on this. I'll merge now and see if I can't write a few tests this afternoon.
